JavaScript development/Frameworks/React.md

```diff
@@ -0,0 +1,598 @@
+- don't use *React.createClass*
+
+``` jsx
+// bad
+export const ComponentName = React.createClass({
+  // ...
+});
+
+// good
+export class ComponentName extends React.Component {
+  // ...
+}
+```
+
+If components are stateless (they don't change state), use **functional components**:
+
+``` jsx
+export function FunctionalComponent({message}) {
+  // ...
+};
+
+// or
+
+export const FunctionalComponent = ({message}) => (
+  // ...
+);
+```
+
+### Formatting
+
+Use self-closing tags, separated with an empty space, for elements that don't have children.
+
+``` jsx
+// bad
+<ChildComponent onClick={this.onClickHandler}></ChildComponent>
+
+// good
+<ChildComponent onClick={this.onClickHandler} />
+```
+
+If there is more than one prop, separate them with line breaks and also, put opening and closing tags in separate lines.
+
+``` jsx
+// bad
+<ChildComponent className={styles.specialStyling}
+  title={this.title}
+  onClick={this.onClickHandler}
+/>
+
+// good
+<ChildComponent
+  className={styles.specialStyling}
+  title={this.title}
+  onClick={this.onClickHandler}
+/>
+
+// good
+<OuterChildComponent
+  className={styles.outerChild}
+  title={this.title}
+  onClick={this.onClickHandler}
+>
+  <InnerChildComponent className={styles.innerChild} />
+</OuterChildComponent>
+```
+
+- always wrap *render* method's return statement in parentheses (it's optional if it's returning a single line of JSX)
+
+``` jsx
+// bad 
+render() {
+  // ...
+  return <div>
+      <ChildComponent />
+    </div>;
+}
+
+// bad 
+render() {
+  // ...
+  return
+    <div>
+      <ChildComponent />
+    </div>;
+}
+
+// good
+render() {
+  // ...
+  return (
+    <div>
+      <ChildComponent />
+    </div>
+  );
+}
+
+// good
+render() {
+  // ...
+  return <ChildComponent />;
+}
+
+// good
+render() {
+  // ...
+  return (
+    <ChildComponent />
+  );
+}
+```
+
+### Passing Props
+Use camelCase for prop names and double quotes for JSX attribute values if you are passing them as a *string*.
+
+``` jsx
+export class ComponentName extends React.Component {
+  // ...
+  render () {
+    return (
+      <ChildComponent
+        defaultTitle="Some Title"
+        title={this.props.title}
+      />
+    );
+  }
+}
+```
+
+Omit values for boolean props if default value is **true**.
+
+``` jsx
+// bad
+<ChildComponent
+  onClick={this.onClickHandler}
+  isPropActive={true}
+/>
+
+// good
+<ChildComponent
+  onClick={this.onClickHandler}
+  isPropActive
+/>
+```
+
+Declare functions before passing them as props instead of passing inline anonymous (arrow) functions. They will be needlessy re-rendered each time, thus effecting performance.
+
+``` jsx
+// bad
+<ChildComponent onClick={() => { /* ... */ } />
+    
+// good
+class ComponentName extends React.Component {
+  onClickHandler() {
+    // ...
+  }
+
+  render () {
+    return (
+      <ChildComponent onClick={this.onClickHandler} />
+    );
+  }
+}
+```
+
+Avoid writing any logic inline. Extract it to the outer scope.
```

Comment:
"outer scope" sounds confusing. Maybe better to change it to something like "outside of the JSX block"